tools/train.py

```diff
@@ -62,6 +68,81 @@ def parse_args():
     return args
 
 
+def determine_max_batch_size(cfg, distributed):
+    def get_fake_input(cfg, orig_img_shape=(128, 128, 3), device='cuda'):
+        test_pipeline = [LoadImage()] + cfg.data.test.pipeline[1:]
+        test_pipeline = Compose(test_pipeline)
+        data = dict(img=np.zeros(orig_img_shape, dtype=np.uint8))
+        data = test_pipeline(data)
+        data = scatter(collate([data], samples_per_gpu=1), [device])[0]
+        return data
+
+    model = build_detector(
+        cfg.model, train_cfg=cfg.train_cfg, test_cfg=cfg.test_cfg).cuda()
+
+    if 'pipeline' in cfg.data.train:
+        img_shape = [t for t in cfg.data.train.pipeline if t['type'] == 'Resize'][0]['img_scale']
+    else:
+        img_shape = [t for t in cfg.data.train.dataset.pipeline if t['type'] == 'Resize'][0][
+            'img_scale']
+
+    channels = 3
+
+    fake_input = get_fake_input(cfg, orig_img_shape=list(img_shape) + [channels])
+    img_shape = fake_input['img_metas'][0][0]['pad_shape']
+
+    width, height = img_shape[0], img_shape[1]
+
+    min_bs = 2
+    max_bs = 512
+    step = 2
+
+    batch_size = min_bs
+    for bs in range(min_bs, max_bs, step):
```

Comment:
sure

Reply:
As it turned out, bisect is worse because in most cases it starts with batch sizes which are significantly bigger than they should be. Memory are not freed after unsuccessful tries as it is expected (even after torch.cuda.empty_cache()) as result we get much smaller resulting batch than it should be.